### ISSUE TITLE
Support updating release notes in metainfo

### ISSUE THREAD

**Issue by vchernin** [flathub-infra/flatpak-external-data-checker#222]
Today f-e-d-c can update the version number and release date in the metainfo/appdata file. I wonder if this could be extended to support adding the changelog details for each update to metainfo/appdata. This would require finding ways to parse an upstream changelog/news file. [Relevant link](https://blog.tenstral.net/2020/03/maintain-release-info-easily-in-metainfo-files.html)

This would ideally improve the situation for many apps on Flathub (eg Chromium) that have a version history like this:
<img src="https://user-images.githubusercontent.com/38842733/130543297-90cb0c56-3aaf-45ae-b6c5-1c5bec3700e6.png" alt="gnome software version info" width="400"/>

**Comment by wjt** [comment]
Yes, that would be great in theory. I fear it may be difficult to build something generic that can parse release information that is intended for end-users rather than developers.

Using Chromium as an example, is there an existing source that could work for this? The best I have found is "Stable Channel Update" posts on https://chromereleases.googleblog.com/ 

The alternative is for the human who merges each PR to add the news by hand.

**Comment by gasinvein** [comment]
Hmm, I as well doubt this could be possible in a generic way. Yet there are sources that provide release info we can grab, like release notes on GitHub. Maybe we should add a GitHubChecker?

**Comment by gasinvein** [comment]
Also, when/if we get either a working sandbox in github actions or a per-app f-e-d-c action, we can allow running custom scripts to retrieve release info during update by f-e-d-c.

**Comment by vchernin** [comment]
See also: https://github.com/ximion/appstream/issues/354

**Comment by nanonyme** [comment]
It would really be best approach to get as many projects as possible ship their appdata files that contain release notes. Core problem of this need to begin with is data that is supposed to come from upstream is getting patched together on Flathub side.

**Comment by vchernin** [comment]
I think perhaps more automated ways to maintain appdata release notes would help upstreams, https://blog.tenstral.net/2020/03/maintain-release-info-easily-in-metainfo-files.html helps but there are so many ways projects write release notes.

**Comment by nanonyme** [comment]
Does it btw support changelog? I noticed mosh actually has NEWS that is abandoned (?) and doesn't have newer releases, newer hings live in a file like https://github.com/mobile-shell/mosh/blob/master/ChangeLog.

**Comment by hadess** [comment]
FlashPrint has a manifest with the latest versions that the software pokes at http://www.ishare3d.com/3dapp/public/FlashPrint-5/appInfo1 which also links to a file with changelogs for the last few versions.

**Comment by gasinvein** [comment]
Finding an upstream changelog is usually easy - converting it to metainfo isn't. How can we parse a text changelog in arbitrary format? I personally doubt that it would be possible without each flatpak app downstream providing its own parser.

**Comment by hadess** [comment]
> I personally doubt that it would be possible without each flatpak app downstream providing its own parser.

I would be fine with that, if updates start happening a bit too often... The main drawback is that you wouldn't want to run random software for the hosted flatpak-external-data-checker, it would need to run locally.

**Comment by nanonyme** [comment]
Would it make sense to patch the release notes in dynamically during app build and only add releases with f-e-d-c? Then the solution would be a lot more flexible. We could use metainfo as side-data for parsing release notes files.

**Comment by gasinvein** [comment]
Hmm, you mean grab upstream changelog with f-e-d-c at check time, commit it with update, and let the app handle it in its own way at built time?

**Comment by nanonyme** [comment]
Well, that is assuming changelog is present as file in release archive

**Comment by gasinvein** [comment]
@hadess Ok, then I suppose we should establish some CLI API to make this language-agnostic.
I can think of two variants:
- After the manifest data is updated, but before metainfo.xml update, a script with well-known name gets executed with release parameters (e.g. version and date) on argv, which must spit out a `<release>` node for given parameters
- Same, but the script gets fed with the current metainfo.xml to stdin with `<release>` already added, and must spit out the whole file, altered the way it likes

**Comment by hadess** [comment]
The former is what would be interesting to me. There's nothing particularly complicated to mangling the upstream changelog, feeding it back into the existing metainfo file is what's more complicated from a casual standpoint.

**Comment by gasinvein** [comment]
@wjt Do you think we can still keep original indentation if we use `<release>` node produced by an external script?

**Comment by wjt** [comment]
Maybe -- the approach would presumably vary depending on whether the stanza to add to the file is treated as raw text or parsed as XML.

Maybe it's not so important.

**Comment by gasinvein** [comment]
@wjt Can we just always re-indent the metainfo.xml on write instead of trying to preserve original indentation? Probably taking editorconfig settings into account.

**Comment by razzeee** [comment]
If we can figure this out, this would IMO be a big win, currently, I end up touching each automatic update to add changelogs again.

**Comment by nanonyme** [comment]
The correct fix is upstream adds release notes into their appstream file. Are you the upstream?

**Comment by razzeee** [comment]
No, not in the projects I have this "problem" with. They all have their appstream data in the flathub repo

**Comment by nanonyme** [comment]
Note that you could also replace appstream file with template and add release notes during app build.

**Comment by razzeee** [comment]
Can we maybe implement something simple optional? Like a configuration, from where to grab the changelog and a regex, that get's used to find the relevant part. Maybe an additional regex to clean up unwanted markdown etc.

IMO we will continue to suffer from waiting for the perfect solution otherwise.

**Comment by dominichayesferen** [comment]
An idea of mine: What if you could also provide a template changelog for cases where the changelog is linked externally for now?

e.g.: "View changes at https://example.com/%VERSION%", which for the new Flatpak version being 0.9.0 would result in "View changes at https://example.com/0.9.0", with other possible translation marks including %RELDATE% for day number of the release, %RELMONTH" for month number of the release and %RELYEAR% for year number of the release

This would be a good stopgap solution until website parsing is potentially added in the future.

**Comment by wjt** [comment]
Good idea. IMO it would be better to use the `<url>` tag within the `<release>` tag: <https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Releases.html>.

I don't think Flathub or GNOME Software actually show these links right now, but they might do if apps used it :)

**Comment by razzeee** [comment]
Flathub shows them

**Comment by dominichayesferen** [comment]
In that case, is it possible for someone to add a way to:
```
    <release version="%VERSION%">
      <url>https://example.com/%VERSION%.html</url>
    </release>
```

and have it be the default release notes addition, with substitutions and automatically inserted date (of the day released)?
Bonus points if the automatically inserted date can be configured to be based on a timezone... unless there's an intended timezone to be used by ALL Flatpaks on Flathub, which case that could do with better communication.

**Comment by connorjclark** [comment]
For a more general solution, would it be reasonable to introduce a callback script that can let us modify the metainfo however we need to? Defer the details to package creators, don't need to support everything in a custom DSL.

**Comment by bbhtt** [comment]
> IMO we will continue to suffer from waiting for the perfect solution otherwise.

This needs to be done correctly or not done at all.

If we start supporting one changelog format from one source even optional, we will get swamped with

- “my special news file produces garbage release notes with x-checker”
- “please support x changelog format from y source”
- ...

and app maintainers will get 

- “your release notes are unintelligible” from users.

The most plausible version of this is allowing some kind of callback script and offload the work away from this tool, for which there exists the more generic issue https://github.com/flathub-infra/flatpak-external-data-checker/issues/252

> In that case, is it possible for someone to add a way to:

See https://github.com/flathub-infra/flatpak-external-data-checker/pull/484
